lib/auth/auth_with_roles.go

```diff
@@ -1483,6 +1483,108 @@ func (a *ServerWithRoles) GetNode(ctx context.Context, namespace, name string) (
 	return node, nil
 }
 
+// ListUnifiedResources returns a paginated list of unified resources filtered by user access.
+func (a *ServerWithRoles) ListUnifiedResources(ctx context.Context, req *proto.ListUnifiedResourcesRequest) (*types.ListResourcesResponse, error) {
+	// Fetch full list of nodes in the backend.
+	startFetch := time.Now()
+	unifiedResources, err := a.authServer.unifiedResourceWatcher.GetUnifiedResources(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	elapsedFetch := time.Since(startFetch)
+
+	// Filter resources to return the ones for the connected identity.
+	filteredResources := make([]types.ResourceWithLabels, 0)
+	startFilter := time.Now()
+	for _, resource := range unifiedResources {
+		switch r := resource.(type) {
+		case types.Server:
+			{
+				if err := a.checkAccessToNode(r); err != nil {
```

Comment:
It looks like only `a.checkAccessToWhateverResourceType(r)` is actually specific to every particular branch. Can we instead declare the `err` variable before the `switch`, use `switch` only to extract `err`, and then perform all the checks separately, after we finish the switch? This will require adding a default branch, where we make sure that if an unknown resource type appears in the future, we don't blindly return it, but I think it should overall harden the code by deduplicating a lot of statements.